Upstream: no, https://bugzilla.gnome.org/show_bug.cgi?id=657346
From e091235c32fe85204673e32174094b18f5f6e102 Mon Sep 17 00:00:00 2001
From: Nikolay Orlyuk <virkony@gmail.com>
Date: Mon, 4 Jun 2012 11:29:16 +0000
Subject: [PATCH] Fix property setter call in SimpleType structs

Current declaration and implementation of setters for SimpleType structs
produces code that expects instance passed by value. But code generater
for call of setter is passing "self" as an pointer to value. Unit test
tests/simpletype.vala helps to reveal that issue.

Changes is from last chunk of patch by Nathan Summers (bug 657346, title
"codegen: Pass SimpleTypes by-val in getter")
---
 codegen/valaccodebasemodule.vala |    4 ++--
 tests/Makefile.am                |    1 +
 tests/simpletype.vala            |   36 ++++++++++++++++++++++++++++++++++++
 3 files changed, 39 insertions(+), 2 deletions(-)
 create mode 100644 tests/simpletype.vala

diff --git a/codegen/valaccodebasemodule.vala b/codegen/valaccodebasemodule.vala
index c158eaf..d0ece90 100644
--- a/codegen/valaccodebasemodule.vala
+++ b/codegen/valaccodebasemodule.vala
@@ -5567,8 +5567,8 @@ public abstract class Vala.CCodeBaseModule : CodeGenerator {
 			/* target instance is first argument */
 			var cinstance = (CCodeExpression) get_ccodenode (instance);
 
-			if (prop.parent_symbol is Struct) {
-				// we need to pass struct instance by reference
+			if (prop.parent_symbol is Struct && !((Struct)prop.parent_symbol).is_simple_type()) {
+				// we need to pass struct instance by reference if it's not a simple type
 				var instance_value = instance.target_value;
 				if (!get_lvalue (instance_value)) {
 					instance_value = store_temp_value (instance_value, instance);
diff --git a/tests/Makefile.am b/tests/Makefile.am
index 9a64c1d..aa14241 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -166,6 +166,7 @@ TESTS = \
 	dbus/bug596862.vala \
 	dbus/bug602003.test \
 	gir/bug651773.test \
+	simpletype.vala \
 	$(NULL)
 
 check-TESTS: $(TESTS)
diff --git a/tests/simpletype.vala b/tests/simpletype.vala
new file mode 100644
index 0000000..87fb8a7
--- /dev/null
+++ b/tests/simpletype.vala
@@ -0,0 +1,36 @@
+public class StaticTable {
+    private static string table[4];
+    public static void init() {
+        table = { "black", "red", "green", "blue" };
+    }
+
+    public static Entry top { get { return 0; } }
+
+    [SimpleType]
+    [CCode (has_type_id = false)]
+    public struct Entry : int {
+        public bool exists { get { return this < table.length; } }
+        public Entry next { get { return this + 1; } }
+        public string data {
+            get {
+                assert (exists);
+                return table[(int)this];
+            }
+            set {
+                assert (exists);
+                table[(int)this] = value;
+            }
+        }
+
+    }
+}
+
+public static void main() {
+    StaticTable.init();
+    var entry = StaticTable.top;
+    assert (entry.exists);
+    assert (entry.data == "black");
+    assert (entry.exists);
+    entry.data = "white";
+    assert (entry.data == "white");
+}
-- 
1.7.10.3

